File: src/Web/Dart/iterator.dart
class CursorIterator<T> {
  CursorIterator(this.values);

  final List<T> values;
  int _index = 0;

  bool get hasNext => _index < values.length;
  T next() => values[_index++];
}

void main() {
  final iterator = CursorIterator<int>([10, 20, 30]);
  final visited = <int>[];
  while (iterator.hasNext) {
    visited.add(iterator.next());
  }

  if (visited.join(',') != '10,20,30' || iterator.hasNext) {
    throw StateError('iterator contract failed');
  }

  print('iterator=${visited.join(',')}');
}

File: src/Scripting/Perl/iterator.pl
use strict;
use warnings;

sub iterator_for {
    my ($items) = @_;
    my $index = 0;
    return sub {
        return undef if $index >= @$items;
        return $items->[$index++];
    };
}

my $next = iterator_for([10, 20, 30]);
my @visited;
while (defined(my $value = $next->())) {
    push @visited, $value;
}
die "iterator contract failed\n" unless join(',', @visited) eq '10,20,30' && !defined($next->());
print "iterator=10,20,30\n";
